File: golang/refactoring-methods/06-method-calls.go
package main

import (
	"fmt"
	"strings"
)

// 42. Renaming a method (Rename Method)
//
// BEFORE: Poorly named method
type CalculatorBefore struct{}

func (c CalculatorBefore) Calc(a, b int) int { // Unclear name
	return a + b
}

// AFTER: Rename method to be more descriptive
type CalculatorAfter struct{}

func (c CalculatorAfter) Add(a, b int) int {
	return a + b
}

// 43. Adding a parameter (Add Parameter)
//
// BEFORE: Method missing required parameter
type EmailSenderBefore struct{}

func (es EmailSenderBefore) SendEmail(to, subject, body string) {
	// Send email with default priority
	priority := "normal"
	fmt.Printf("Sending email to %s with priority %s\n", to, priority)
}

// AFTER: Add parameter
type EmailSenderAfter struct{}

func (es EmailSenderAfter) SendEmail(to, subject, body, priority string) {
	if priority == "" {
		priority = "normal"
	}
	fmt.Printf("Sending email to %s with priority %s\n", to, priority)
}

// 44. Deleting a parameter (Remove Parameter)
//
// BEFORE: Unnecessary parameter
type ReportGeneratorBefore struct{}

func (rg ReportGeneratorBefore) GenerateReport(data string, format string, includeHeader bool) {
	if format == "html" {
		// Always include header for HTML
		includeHeader = true
	}
	fmt.Printf("Generating %s report with header: %v\n", format, includeHeader)
}

// AFTER: Remove unnecessary parameter
type ReportGeneratorAfter struct{}

func (rg ReportGeneratorAfter) GenerateReport(data, format string) {
	includeHeader := (format == "html")
	fmt.Printf("Generating %s report with header: %v\n", format, includeHeader)
}

// 45. Separation of Query and Modifier (Separate Query from Modifier)
//
// BEFORE: Method that both queries and modifies
type BankAccountBefore struct {
	balance float64
}

func (ba *BankAccountBefore) Withdraw(amount float64) bool {
	if ba.balance >= amount {
		ba.balance -= amount
		return true
	}
	return false
}

// AFTER: Separate query from modifier
type BankAccountAfter struct {
	balance float64
}

func (ba BankAccountAfter) CanWithdraw(amount float64) bool {
	return ba.balance >= amount
}

func (ba *BankAccountAfter) Withdraw(amount float64) bool {
	if ba.CanWithdraw(amount) {
		ba.balance -= amount
		return true
	}
	return false
}

// 46. Parameterization of the method (Parameterize Method)
//
// BEFORE: Similar methods with different values
type ReportGeneratorParamBefore struct{}

func (rg ReportGeneratorParamBefore) GenerateWeeklyReport() {
	rg.generateReport(7)
}

func (rg ReportGeneratorParamBefore) GenerateMonthlyReport() {
	rg.generateReport(30)
}

func (rg ReportGeneratorParamBefore) GenerateQuarterlyReport() {
	rg.generateReport(90)
}

func (rg ReportGeneratorParamBefore) generateReport(days int) {
	fmt.Printf("Generating report for %d days\n", days)
}

// AFTER: Parameterize method
type ReportGeneratorParamAfter struct{}

func (rg ReportGeneratorParamAfter) GenerateReport(days int) {
	fmt.Printf("Generating report for %d days\n", days)
}

func (rg ReportGeneratorParamAfter) GenerateWeeklyReport() {
	rg.GenerateReport(7)
}

func (rg ReportGeneratorParamAfter) GenerateMonthlyReport() {
	rg.GenerateReport(30)
}

func (rg ReportGeneratorParamAfter) GenerateQuarterlyReport() {
	rg.GenerateReport(90)
}

// 47. Replacing a parameter with explicit methods (Replace Parameter with Explicit Methods)
//
// BEFORE: Parameter determines behavior
type EmployeeBefore struct {
	name   string
	salary float64
}

func (e EmployeeBefore) CalculateBonus(bonusType string) float64 {
	switch bonusType {
	case "performance":
		return e.salary * 0.1
	case "yearly":
		return e.salary * 0.05
	case "special":
		return e.salary * 0.15
	default:
		return 0
	}
}

// AFTER: Replace parameter with explicit methods
type EmployeeAfter struct {
	name   string
	salary float64
}

func (e EmployeeAfter) CalculatePerformanceBonus() float64 {
	return e.salary * 0.1
}

func (e EmployeeAfter) CalculateYearlyBonus() float64 {
	return e.salary * 0.05
}

func (e EmployeeAfter) CalculateSpecialBonus() float64 {
	return e.salary * 0.15
}

// 48. Save the Whole Object
//
// BEFORE: Passing individual fields
type OrderItemBefore struct {
	name     string
	price    float64
	quantity int
}

type OrderProcessorBefore struct{}

func (op OrderProcessorBefore) CalculateItemTotal(name string, price float64, quantity int) float64 {
	return price * float64(quantity)
}

// AFTER: Save the whole object
type OrderItemAfter struct {
	name     string
	price    float64
	quantity int
}

type OrderProcessorAfter struct{}

func (op OrderProcessorAfter) CalculateItemTotal(item OrderItemAfter) float64 {
	return item.price * float64(item.quantity)
}

// 49. Replacing a parameter with a method call (Replace Parameter with Method)
//
// BEFORE: Parameter passed that could be calculated
type CustomerBefore struct {
	name  string
	level string
}

type DiscountCalculatorBefore struct{}

func (dc DiscountCalculatorBefore) CalculateDiscount(customer CustomerBefore, purchaseAmount float64) float64 {
	discountRate := 0.0
	switch customer.level {
	case "gold":
		discountRate = 0.1
	case "silver":
		discountRate = 0.05
	case "bronze":
		discountRate = 0.02
	}
	return purchaseAmount * discountRate
}

// AFTER: Replace parameter with method call
type CustomerAfter struct {
	name  string
	level string
}

func (c CustomerAfter) GetDiscountRate() float64 {
	switch c.level {
	case "gold":
		return 0.1
	case "silver":
		return 0.05
	case "bronze":
		return 0.02
	default:
		return 0.0
	}
}

type DiscountCalculatorAfter struct{}

func (dc DiscountCalculatorAfter) CalculateDiscount(customer CustomerAfter, purchaseAmount float64) float64 {
	return purchaseAmount * customer.GetDiscountRate()
}

// 50. Introduction of the boundary object (Introduce Parameter Object)
//
// BEFORE: Multiple parameters
type ReservationServiceBefore struct{}

func (rs ReservationServiceBefore) MakeReservation(date, startTime, endTime, customerName, customerEmail string, partySize int) {
	fmt.Printf("Reservation for %s (%s) on %s from %s to %s for %d people\n",
		customerName, customerEmail, date, startTime, endTime, partySize)
}

// AFTER: Introduce parameter object
type ReservationDetails struct {
	Date          string
	StartTime     string
	EndTime       string
	CustomerName  string
	CustomerEmail string
	PartySize     int
}

type ReservationServiceAfter struct{}

func (rs ReservationServiceAfter) MakeReservation(details ReservationDetails) {
	fmt.Printf("Reservation for %s (%s) on %s from %s to %s for %d people\n",
		details.CustomerName, details.CustomerEmail, details.Date,
		details.StartTime, details.EndTime, details.PartySize)
}

// 51. Removing the Value Setting Method (Remove Setting Method)
//
// BEFORE: Setter that shouldn't exist
type ImmutableObjectBefore struct {
	value string
}

func (io *ImmutableObjectBefore) SetValue(value string) {
	io.value = value // This breaks immutability
}

func (io ImmutableObjectBefore) GetValue() string {
	return io.value
}

// AFTER: Remove setting method
type ImmutableObjectAfter struct {
	value string
}

func NewImmutableObjectAfter(value string) *ImmutableObjectAfter {
	return &ImmutableObjectAfter{value: value}
}

func (io ImmutableObjectAfter) GetValue() string {
	return io.value
}

// 52. Hiding a method (Hide Method)
//
// BEFORE: Public method that should be private
type DataProcessorBefore struct{}

func (dp DataProcessorBefore) ValidateInput(input string) bool {
	return len(strings.TrimSpace(input)) > 0
}

func (dp DataProcessorBefore) ProcessInput(input string) string {
	if dp.ValidateInput(input) {
		return strings.ToUpper(input)
	}
	return ""
}

// AFTER: Hide method
type DataProcessorAfter struct{}

func (dp DataProcessorAfter) ProcessInput(input string) string {
	if dp.validateInput(input) {
		return strings.ToUpper(input)
	}
	return ""
}

func (dp DataProcessorAfter) validateInput(input string) bool {
	return len(strings.TrimSpace(input)) > 0
}

// 53. Replacing the constructor with the factory method (Replace Constructor with Factory Method)
//
// BEFORE: Constructor with complex logic
type DatabaseConnectionBefore struct {
	host     string
	port     int
	database string
}

func NewDatabaseConnectionBefore(connectionString string) *DatabaseConnectionBefore {
	// Complex parsing logic in constructor
	parts := strings.Split(connectionString, ":")
	host := parts[0]
	port := 5432 // default
	database := "default"

	if len(parts) > 1 {
		// Parse port
		fmt.Sscanf(parts[1], "%d", &port)
	}
	if len(parts) > 2 {
		database = parts[2]
	}

	return &DatabaseConnectionBefore{
		host:     host,
		port:     port,
		database: database,
	}
}

// AFTER: Replace constructor with factory method
type DatabaseConnectionAfter struct {
	host     string
	port     int
	database string
}

func NewDatabaseConnectionAfter(connectionString string) *DatabaseConnectionAfter {
	return CreateDatabaseConnection(connectionString)
}

func CreateDatabaseConnection(connectionString string) *DatabaseConnectionAfter {
	// Complex parsing logic in factory method
	parts := strings.Split(connectionString, ":")
	host := parts[0]
	port := 5432 // default
	database := "default"

	if len(parts) > 1 {
		// Parse port
		fmt.Sscanf(parts[1], "%d", &port)
	}
	if len(parts) > 2 {
		database = parts[2]
	}

	return &DatabaseConnectionAfter{
		host:     host,
		port:     port,
		database: database,
	}
}

// Example usage for method calls
func demonstrateMethodCalls() {
	fmt.Println("=== 42. Rename Method ===")

	calcBefore := CalculatorBefore{}
	calcAfter := CalculatorAfter{}

	fmt.Printf("Result before: %d\n", calcBefore.Calc(5, 3))
	fmt.Printf("Result after: %d\n", calcAfter.Add(5, 3))

	fmt.Println("\n=== 43. Add Parameter ===")

	senderBefore := EmailSenderBefore{}
	senderAfter := EmailSenderAfter{}

	senderBefore.SendEmail("user@example.com", "Test", "Body")
	senderAfter.SendEmail("user@example.com", "Test", "Body", "high")

	fmt.Println("\n=== 44. Remove Parameter ===")

	reportBefore := ReportGeneratorBefore{}
	reportAfter := ReportGeneratorAfter{}

	reportBefore.GenerateReport("data", "html", false) // includeHeader ignored for HTML
	reportAfter.GenerateReport("data", "html")         // No unnecessary parameter

	fmt.Println("\n=== 45. Separate Query from Modifier ===")

	accountBefore := &BankAccountBefore{balance: 1000.0}
	accountAfter := &BankAccountAfter{balance: 1000.0}

	fmt.Printf("Can withdraw $500 before: %v\n", accountBefore.balance >= 500)
	successBefore := accountBefore.Withdraw(500)
	fmt.Printf("Withdrawal success before: %v, Balance: $%.2f\n", successBefore, accountBefore.balance)

	fmt.Printf("Can withdraw $500 after: %v\n", accountAfter.CanWithdraw(500))
	successAfter := accountAfter.Withdraw(500)
	fmt.Printf("Withdrawal success after: %v, Balance: $%.2f\n", successAfter, accountAfter.balance)

	fmt.Println("\n=== 46. Parameterize Method ===")

	genBefore := ReportGeneratorParamBefore{}
	genAfter := ReportGeneratorParamAfter{}

	fmt.Print("Weekly report before: ")
	genBefore.GenerateWeeklyReport()

	fmt.Print("Weekly report after: ")
	genAfter.GenerateWeeklyReport()

	fmt.Println("\n=== 47. Replace Parameter with Explicit Methods ===")

	employeeBefore := EmployeeBefore{name: "John", salary: 50000}
	employeeAfter := EmployeeAfter{name: "John", salary: 50000}

	fmt.Printf("Performance bonus before: $%.2f\n", employeeBefore.CalculateBonus("performance"))
	fmt.Printf("Performance bonus after: $%.2f\n", employeeAfter.CalculatePerformanceBonus())

	fmt.Println("\n=== 48. Preserve Whole Object ===")

	processorBefore := OrderProcessorBefore{}
	processorAfter := OrderProcessorAfter{}

	itemBefore := OrderItemBefore{name: "Widget", price: 10.0, quantity: 5}
	itemAfter := OrderItemAfter{name: "Widget", price: 10.0, quantity: 5}

	fmt.Printf("Total before: $%.2f\n", processorBefore.CalculateItemTotal(itemBefore.name, itemBefore.price, itemBefore.quantity))
	fmt.Printf("Total after: $%.2f\n", processorAfter.CalculateItemTotal(itemAfter))

	fmt.Println("\n=== 50. Introduce Parameter Object ===")

	serviceBefore := ReservationServiceBefore{}
	serviceAfter := ReservationServiceAfter{}

	// Before - many parameters
	serviceBefore.MakeReservation("2024-01-15", "19:00", "21:00", "John Doe", "john@example.com", 4)

	// After - single parameter object
	details := ReservationDetails{
		Date:          "2024-01-15",
		StartTime:     "19:00",
		EndTime:       "21:00",
		CustomerName:  "John Doe",
		CustomerEmail: "john@example.com",
		PartySize:     4,
	}
	serviceAfter.MakeReservation(details)
}

File: python/refactoring-methods/06-method-calls.py
# 42. Renaming a method (Rename Method)
#
# BEFORE: Poorly named method
class CalculatorBefore:
    def calc(self, a, b):  # Unclear name
        return a + b

# AFTER: Rename method to be more descriptive
class CalculatorAfter:
    def add(self, a, b):
        return a + b

# 43. Adding a parameter (Add Parameter)
#
# BEFORE: Method missing required parameter
class EmailSenderBefore:
    def send_email(self, to_email, subject, body):
        # Send email with default priority
        priority = 'normal'
        # Send logic
        pass

# AFTER: Add parameter
class EmailSenderAfter:
    def send_email(self, to_email, subject, body, priority='normal'):
        # Send logic with priority
        pass

# 44. Deleting a parameter (Remove Parameter)
#
# BEFORE: Unnecessary parameter
class ReportGeneratorBefore:
    def generate_report(self, data, report_format, include_header=True):
        if report_format == 'html':
            # Always include header for HTML
            include_header = True
        # Generate report
        pass

# AFTER: Remove unnecessary parameter
class ReportGeneratorAfter:
    def generate_report(self, data, report_format):
        include_header = (report_format == 'html')
        # Generate report
        pass

# 45. Separation of Query and Modifier (Separate Query from Modifier)
#
# BEFORE: Method that both queries and modifies
class BankAccountBefore:
    def __init__(self):
        self._balance = 0

    def withdraw(self, amount):
        if self._balance >= amount:
            self._balance -= amount
            return True
        return False

# AFTER: Separate query from modifier
class BankAccountAfter:
    def __init__(self):
        self._balance = 0

    def can_withdraw(self, amount):
        return self._balance >= amount

    def withdraw(self, amount):
        if self.can_withdraw(amount):
            self._balance -= amount
            return True
        return False

# 46. Parameterization of the method (Parameterize Method)
#
# BEFORE: Similar methods with different values
class ReportGeneratorParamBefore:
    def generate_weekly_report(self):
        return self.generate_report(7)

    def generate_monthly_report(self):
        return self.generate_report(30)

    def generate_quarterly_report(self):
        return self.generate_report(90)

    def generate_report(self, days):
        # Generate report for specified days
        pass

# AFTER: Parameterize method
class ReportGeneratorParamAfter:
    def generate_report(self, days):
        # Generate report for specified days
        pass

    def generate_weekly_report(self):
        return self.generate_report(7)

    def generate_monthly_report(self):
        return self.generate_report(30)

    def generate_quarterly_report(self):
        return self.generate_report(90)

# 47. Replacing a parameter with explicit methods (Replace Parameter with Explicit Methods)
#
# BEFORE: Parameter determines behavior
class EmployeeExplicitBefore:
    ENGINEER = 0
    SALESMAN = 1
    MANAGER = 2

    def __init__(self, employee_type):
        self._type = employee_type

    def get_salary(self, base_salary):
        if self._type == self.ENGINEER:
            return base_salary * 1.0
        elif self._type == self.SALESMAN:
            return base_salary * 1.1
        elif self._type == self.MANAGER:
            return base_salary * 1.2
        else:
            return base_salary

# AFTER: Replace parameter with explicit methods
class EmployeeExplicitAfter:
    def get_engineer_salary(self, base_salary):
        return base_salary * 1.0

    def get_salesman_salary(self, base_salary):
        return base_salary * 1.1

    def get_manager_salary(self, base_salary):
        return base_salary * 1.2

# 48. Save the Whole Object
#
# BEFORE: Passing individual fields
class OrderWholeBefore:
    def __init__(self, customer_name, customer_address):
        self._customer = {'name': customer_name, 'address': customer_address}

    def calculate_shipping(self):
        return self._get_shipping_cost(self._customer['name'], self._customer['address'])

    def _get_shipping_cost(self, name, address):
        # Calculate based on name and address
        return 10.0

# AFTER: Pass whole object
class Customer:
    def __init__(self, name, address):
        self._name = name
        self._address = address

    def get_name(self):
        return self._name

    def get_address(self):
        return self._address

class OrderWholeAfter:
    def __init__(self, customer):
        self._customer = customer

    def calculate_shipping(self):
        return self._get_shipping_cost(self._customer)

    def _get_shipping_cost(self, customer):
        # Calculate based on customer object
        return 10.0

# 49. Replacing a parameter with a method call (Replace Parameter with Method)
#
# BEFORE: Parameter calculated outside method
class DiscountCalculatorParamBefore:
    def calculate_discount(self, price, customer_type):
        # customer_type passed in
        return price * self._get_discount_rate(customer_type)

    def _get_discount_rate(self, customer_type):
        if customer_type == 'premium':
            return 0.1
        elif customer_type == 'regular':
            return 0.05
        else:
            return 0.0

class OrderParamBefore:
    def __init__(self, customer):
        self._customer = customer

    def get_discounted_price(self, price):
        calculator = DiscountCalculatorParamBefore()
        return calculator.calculate_discount(price, self._customer.get_type())

# AFTER: Replace parameter with method call
class DiscountCalculatorParamAfter:
    def calculate_discount(self, price, customer):
        return price * self._get_discount_rate(customer.get_type())

    def _get_discount_rate(self, customer_type):
        if customer_type == 'premium':
            return 0.1
        elif customer_type == 'regular':
            return 0.05
        else:
            return 0.0

class OrderParamAfter:
    def __init__(self, customer):
        self._customer = customer

    def get_discounted_price(self, price):
        calculator = DiscountCalculatorParamAfter()
        return calculator.calculate_discount(price, self._customer)

# 50. Introduction of the boundary object (Introduce Parameter Object)
#
# BEFORE: Multiple parameters
class TemperatureRangeBefore:
    def within_range(self, min_temp, max_temp, current_temp):
        return current_temp >= min_temp and current_temp <= max_temp

    def get_average_temp(self, min_temp, max_temp):
        return (min_temp + max_temp) / 2

# AFTER: Introduce parameter object
class TemperatureRange:
    def __init__(self, min_temp, max_temp):
        self._min_temp = min_temp
        self._max_temp = max_temp

    def get_min_temp(self):
        return self._min_temp

    def get_max_temp(self):
        return self._max_temp

    def within_range(self, current_temp):
        return current_temp >= self._min_temp and current_temp <= self._max_temp

    def get_average_temp(self):
        return (self._min_temp + self._max_temp) / 2

# 51. Removing the Value Setting Method
#
# BEFORE: Setter that's not needed
class SensorBefore:
    def __init__(self, temperature):
        self._temperature = temperature

    def get_temperature(self):
        return self._temperature

    def set_temperature(self, temperature):  # Not needed if immutable
        self._temperature = temperature

# AFTER: Remove setter for immutable object
class SensorAfter:
    def __init__(self, temperature):
        self._temperature = temperature

    def get_temperature(self):
        return self._temperature

    # set_temperature removed

# 52. Hiding a method (Hide Method)
#
# BEFORE: Public method that should be private
class DataProcessorHideBefore:
    def validate_data(self, data):  # Should be private
        return data and isinstance(data, list)

    def process_data(self, data):
        if self.validate_data(data):
            # Process data
            pass

# AFTER: Hide method
class DataProcessorHideAfter:
    def _validate_data(self, data):
        return data and isinstance(data, list)

    def process_data(self, data):
        if self._validate_data(data):
            # Process data
            pass

# 53. Replacing the constructor with the factory method (Replace Constructor with Factory Method)
#
# BEFORE: Complex constructor
class ComplexObjectBefore:
    def __init__(self, obj_type, config=None):
        if config is None:
            config = {}
        self._type = obj_type
        self._config = config

        if obj_type == 'database':
            self._config.update({'host': 'localhost', 'port': 3306})
        elif obj_type == 'file':
            self._config.update({'path': '/tmp', 'format': 'json'})

# AFTER: Replace constructor with factory method
class ComplexObjectAfter:
    def __init__(self, obj_type, config):
        self._type = obj_type
        self._config = config

    @staticmethod
    def create_database_connection(config=None):
        if config is None:
            config = {}
        final_config = {'host': 'localhost', 'port': 3306}
        final_config.update(config)
        return ComplexObjectAfter('database', final_config)

    @staticmethod
    def create_file_handler(config=None):
        if config is None:
            config = {}
        final_config = {'path': '/tmp', 'format': 'json'}
        final_config.update(config)
        return ComplexObjectAfter('file', final_config)

# 54. Encapsulation of top-down type conversion (Encapsulate Downcast)
#
# BEFORE: Downcast in client code
class ShapeCollectionBefore:
    def __init__(self):
        self._shapes = []

    def add_shape(self, shape):
        self._shapes.append(shape)

    def get_shapes(self):
        return self._shapes

# Client code would do:
# circles = [shape for shape in collection.get_shapes() if isinstance(shape, Circle)]

# AFTER: Encapsulate downcast
class ShapeCollectionAfter:
    def __init__(self):
        self._shapes = []

    def add_shape(self, shape):
        self._shapes.append(shape)

    def get_circles(self):
        return [shape for shape in self._shapes if isinstance(shape, Circle)]

    def get_squares(self):
        return [shape for shape in self._shapes if isinstance(shape, Square)]

# Placeholder classes for the example
class Circle:
    pass

class Square:
    pass

# 55. Replacing the error code with an exceptional situation (Replace Error Code with Exception)
#
# BEFORE: Error codes
class FileReaderErrorBefore:
    FILE_NOT_FOUND = 1
    PERMISSION_DENIED = 2

    def read_file(self, filename):
        if not os.path.exists(filename):
            return self.FILE_NOT_FOUND

        if not os.access(filename, os.R_OK):
            return self.PERMISSION_DENIED

        with open(filename, 'r') as f:
            return f.read()

# Client code would check for error codes

# AFTER: Replace error codes with exceptions
class FileNotFoundException(Exception):
    pass

class PermissionDeniedException(Exception):
    pass

class FileReaderExceptionAfter:
    def read_file(self, filename):
        if not os.path.exists(filename):
            raise FileNotFoundException(f"File not found: {filename}")

        if not os.access(filename, os.R_OK):
            raise PermissionDeniedException(f"Permission denied: {filename}")

        with open(filename, 'r') as f:
            return f.read()

# Client code would use try/except

# 56. Replacing an exceptional situation with a check (Replace Exception with Test)
#
# BEFORE: Using exception for control flow
class EmptyStackException(Exception):
    pass

class StackExceptionBefore:
    def __init__(self):
        self._items = []

    def pop(self):
        if not self._items:
            raise EmptyStackException()
        return self._items.pop()

# Client code would catch exception

# AFTER: Replace exception with test
class StackTestAfter:
    def __init__(self):
        self._items = []

    def is_empty(self):
        return not self._items

    def pop(self):
        return self._items.pop()

# Client code would check is_empty() first

import os
